CadRevealFbxProvider.Tests/BatchUtils/GeometryInstancerTests.cs

```diff
@@ -0,0 +1,347 @@
+            Matrix4x4.Identity, 1, new Color(), new BoundingBox(new Vector3(), new Vector3())));
+        primitives.Add(new InstancedMesh(instanceIds[0], new Mesh(dummyVertices.ToArray(), dummyIndices.ToArray(), 0.01f),
+            Matrix4x4.Identity, 1, new Color(), new BoundingBox(new Vector3(), new Vector3())));
+        primitives.Add(new TriangleMesh(new Mesh(dummyVertices.ToArray(), dummyIndices.ToArray(), 0.01f),
+            1, new Color(), new BoundingBox(new Vector3(), new Vector3())));
+        primitives.Add(new TriangleMesh(new Mesh(dummyVertices.ToArray(), dummyIndices.ToArray(), 0.01f),
+            1, new Color(), new BoundingBox(new Vector3(), new Vector3())));
+        primitives.Add(new InstancedMesh(instanceIds[1], new Mesh(dummyVertices.ToArray(), dummyIndices.ToArray(), 0.02f),
+            Matrix4x4.Identity, 1, new Color(), new BoundingBox(new Vector3(), new Vector3())));
+        primitives.Add(new InstancedMesh(instanceIds[1], new Mesh(dummyVertices.ToArray(), dummyIndices.ToArray(), 0.02f),
+            Matrix4x4.Identity, 1, new Color(), new BoundingBox(new Vector3(), new Vector3())));
+        primitives.Add(new InstancedMesh(instanceIds[1], new Mesh(dummyVertices.ToArray(), dummyIndices.ToArray(), 0.02f),
+            Matrix4x4.Identity, 1, new Color(), new BoundingBox(new Vector3(), new Vector3())));
+
+        node.Geometries = primitives.ToArray();
+
+        return new APrimitive[] { primitives[3], primitives[7] }; // Primitives to be instanced
+    }
+
+    private void CheckNodePattern1(CadRevealNode node)
+    {
+        // Check that the non-instanced objects are still all distinct
+        for (var i = 1; i < node.Geometries.Length; i++)
+        {
+            AssertMeshReferences(node.Geometries[0], node.Geometries[i], false);
+        }
+        for (var i = 2; i < node.Geometries.Length; i++)
+        {
+            AssertMeshReferences(node.Geometries[1], node.Geometries[i], false);
+        }
+        for (var i = 2; i < node.Geometries.Length; i++)
+        {
+            if (i != 5) AssertMeshReferences(node.Geometries[5], node.Geometries[i], false);
+        }
+        for (var i = 3; i < node.Geometries.Length; i++)
+        {
+            if (i != 6) AssertMeshReferences(node.Geometries[6], node.Geometries[i], false);
+        }
+
+        // Check that the instanced objects with same instance ID now share the same meshes
+        for (var i = 3; i <= 4; i++)
+        {
+            AssertMeshReferences(node.Geometries[2], node.Geometries[i], true);
+        }
+    }
+
+    private void CheckNodePattern2(CadRevealNode node)
+    {
+        // Check that the non-instanced objects are still all distinct
+        for (var i = 1; i < node.Geometries.Length; i++)
+        {
+            AssertMeshReferences(node.Geometries[0], node.Geometries[i], false);
+        }
+        for (var i = 2; i < node.Geometries.Length; i++)
```

Comment:
The offset is there to prevent comparing the same thing twice (e.g., a_i==a_(i+1) and a_(i+1)==a_i). However, we may be able to reduce this to nested loops instead of consecutive loops, and it really doesn't matter if we do compare the same items twice.

Reply:
I would prefer readability over redundant checks, but OK to keep as is if you want